### PR TITLE
Typeorm Upgrade

### PR DESCRIPTION
**Pull request by valentinyanakiev**
A few major notes:
   - Transitive eager loaded dependencies below 3rd level (A --> B --> C --> D), B will eagerly load C, but C won't eager load D. Added explicit loading of those entities on all places we found in the test-suites repo.
   - .create invokes the default constructor of the `Entity`. I have altered some definitions to match the database - mandatory (not null) fields in the DB can (and should) be mandatory in the entities as well, and then they are not required in the constructor.
   - FindByIds is replaced with `FindBy`, `where` and `In`
   - Find + where works differently, the whole `where` is inside the `options` now
   - Typeorm introduces the concept of a `DataSource` now. It is used by the migrations, migrations altogether use different parameters
   - `OrmConfig` concept in TypeORM is dropped
   - To create a new migration, run `npm run migration:run src/migrations/[name_of_migration]` - this is slightly different to the behavior before
   - `DataSources` use named connections. The default name is `default`. It needs to be provided when the new @InjectEntityManager is used. It simplifies usage of repositories in scenarios where we want to get any repository generically. 
   - There are a lot of deprecated methods...
   - I have added `| never` when we query data and throw an exception
   - `undefined` is largely replaced as `null` as return type
   - `findOne` default behavior is changed now - if we pass no parameters it doesn't return the first entry of the table, it returns `undefined`

### CODE REVIEW

**Comment by Comoque1** [comment]
Previously, when creating an `actorGroup`, without actors, the attribute `actors` where `null` now it is `[ ]`
image

Also when querying created `actors`, the `actorGroup` returns null
image

**Comment by Comoque1** [comment]
Aspect template tagsets are returning `null` instead of the actual value
image

**Comment by Comoque1** [comment]
Querying group parent by organization type, returns, community type:
image

**Comment by valentinyanakiev** [comment]
> Aspect template tagsets are returning `null` instead of the actual value image

that is fixed now

**Comment by valentinyanakiev** [comment]
> Previously, when creating an `actorGroup`, without actors, the attribute `actors` where `null` now it is `[ ]` image
> 
> Also when querying created `actors`, the `actorGroup` returns null image

fixed

**Comment by Comoque1** [comment]
Community Admin doesn't receive notification, when the preference is enabled.

File: src/domain/agent/credential/credential.entity.ts [hunk 1]
```diff
@@ -18,10 +18,7 @@ export class Credential extends BaseAlkemioEntity implements ICredential {
   })
   agent?: Agent;
 
-  constructor(type: string, resourceID: string) {
+  constructor() {
```
Comment (hero101) [inline src/domain/agent/credential/credential.entity.ts:21]: why was this required?

File: src/domain/agent/credential/credential.service.ts [hunk 2]
```diff
@@ -19,13 +19,13 @@ export class CredentialService {
...
   ): Promise<ICredential> {
-    const credential = Credential.create(credentialInput);
+    const credential = Credential.create({ ...credentialInput });
     await this.credentialRepository.save(credential);
     return credential;
   }
 
   async getCredentialOrFail(credentialID: string): Promise<ICredential> {
-    const credential = await this.credentialRepository.findOne({
+    const credential = await this.credentialRepository.findOneBy({
```
Comment (hero101) [inline src/domain/agent/credential/credential.service.ts:28]: in agent.service.ts findOne was used instead - which change is the correct one or it's was because of the provided options?

File: src/domain/challenge/challenge/challenge.service.ts [hunk 3]
```diff
@@ -601,24 +609,24 @@ export class ChallengeService {
...
-    }[] = await getConnection().query(sqlQuery);
+    }[] = await this.entityManager.connection.query(sqlQuery);
 
     return queryResult.challengesCount;
   }
 
   async getChildChallengesCount(challengeID: string): Promise<number> {
-    return await this.challengeRepository.count({
-      where: { parentChallenge: challengeID },
+    return await this.challengeRepository.countBy({
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.service.ts:626]: why

File: src/domain/collaboration/agreement/agreement.entity.ts [hunk 4]
```diff
@@ -26,9 +23,8 @@ export class Agreement extends BaseAlkemioEntity implements IAgreement {
   @JoinColumn()
   tagset: Tagset;
 
-  constructor(name: string) {
+  constructor() {
     super();
-    this.name = name;
-    this.tagset = new Tagset(RestrictedTagsetNames.DEFAULT);
+    this.tagset = new Tagset();
```
Comment (hero101) [inline src/domain/collaboration/agreement/agreement.entity.ts:28]: this is possibly changing behaviour

File: src/domain/collaboration/aspect/aspect.service.ts [hunk 5]
```diff
@@ -57,7 +57,7 @@ export class AspectService {
   public async deleteAspect(deleteData: DeleteAspectInput): Promise<IAspect> {
     const aspectID = deleteData.ID;
     const aspect = await this.getAspectOrFail(aspectID, {
-      relations: ['profile'],
+      relations: ['profile', 'profile.tagset'],
```
Comment (hero101) [inline src/domain/collaboration/aspect/aspect.service.ts]: i don't see tagset used in this method

File: src/domain/collaboration/aspect/aspect.service.ts [hunk 6]
```diff
@@ -132,7 +132,7 @@ export class AspectService {
 
   public async getCardProfile(aspect: IAspect): Promise<ICardProfile> {
     const aspectLoaded = await this.getAspectOrFail(aspect.id, {
-      relations: ['profile'],
+      relations: ['profile', 'profile.tagset'],
```
Comment (hero101) [inline src/domain/collaboration/aspect/aspect.service.ts]: the proper way to do this imo is to provide either options or the relation param - the method is loading the profile from a card, the tagset is used just in one use-case where the method is used, the other use-cases probably won't need the tagset loaded here.

File: src/domain/collaboration/callout/callout.service.ts [hunk 7]
```diff
@@ -450,7 +450,7 @@ export class CalloutService {
     shuffle?: boolean
   ): Promise<IAspect[]> {
     const loadedCallout = await this.getCalloutOrFail(callout.id, {
-      relations: ['aspects'],
+      relations: ['aspects', 'aspects.comments'],
```
Comment (hero101) [inline src/domain/collaboration/callout/callout.service.ts]: comments are not needed for the impl of this method, a use-case is requiring the comments to be loaded

File: src/domain/common/preference-set/preference.set.service.ts [hunk 8]
```diff
@@ -66,11 +66,11 @@ export class PreferenceSetService {
   async getPreferenceSetOrFail(
     preferenceSetID: string,
     options?: FindOneOptions<PreferenceSet>
-  ): Promise<IPreferenceSet> {
-    const preferenceSet = await PreferenceSet.findOne(
-      { id: preferenceSetID },
-      options
-    );
+  ): Promise<IPreferenceSet | never> {
+    const preferenceSet = await PreferenceSet.findOne({
```
Comment (hero101) [inline src/domain/common/preference-set/preference.set.service.ts:70]: in nvp.service.ts findOne is changed with findOneBy, here it isn't- why?

File: src/domain/community/community/community.service.ts [hunk 9]
```diff
@@ -420,7 +420,7 @@ export class CommunityService {
 
   async getCommunication(communityID: string): Promise<ICommunication> {
     const community = await this.getCommunityOrFail(communityID, {
-      relations: ['communication'],
+      relations: ['communication', 'communication.updates'],
```
Comment (hero101) [inline src/domain/community/community/community.service.ts]: the imp is not dependant on updates

File: src/domain/community/organization/organization.service.ts [hunk 10]
```diff
@@ -216,6 +216,7 @@ export class OrganizationService {
         'agent',
         'verification',
         'preferenceSet',
+        'preferenceSet.preferences',
```
Comment (hero101) [inline src/domain/community/organization/organization.service.ts]: where is this used

Comment (hero101) [inline src/domain/community/organization/organization.service.ts]: where is this used

File: src/domain/community/organization/organization.service.ts [hunk 11]
```diff
@@ -447,7 +448,7 @@ export class OrganizationService {
     );
     // Try to find the organization
     const organization = await this.getOrganizationOrFail(orgID, {
-      relations: ['groups'],
+      relations: ['groups', 'groups.profile'],
```
Comment (hero101) [inline src/domain/community/organization/organization.service.ts:450]: not used in the impl

File: src/domain/community/profile/profile.service.ts [hunk 12]
```diff
@@ -208,8 +208,11 @@ export class ProfileService {
   async getProfileOrFail(
     profileID: string,
     options?: FindOneOptions<Profile>
-  ): Promise<IProfile> {
-    const profile = await Profile.findOne({ id: profileID }, options);
+  ): Promise<IProfile | never> {
+    const profile = await Profile.findOne({
+      where: { id: profileID },
+      ...options,
```
Comment (hero101) [inline src/domain/community/profile/profile.service.ts:213]: the where in options will override the where

File: src/domain/community/user/user.service.ts [hunk 13]
```diff
@@ -284,7 +284,12 @@ export class UserService {
   async deleteUser(deleteData: DeleteUserInput): Promise<IUser> {
     const userID = deleteData.ID;
     const user = await this.getUserOrFail(userID, {
-      relations: ['profile', 'agent', 'preferenceSet'],
+      relations: [
+        'profile',
+        'agent',
+        'preferenceSet',
+        'preferenceSet.preferences',
```
Comment (hero101) [inline src/domain/community/user/user.service.ts]: where is this used

File: src/library/library/library.service.ts [hunk 14]
```diff
@@ -24,7 +24,7 @@ export class LibraryService {
   ) {}
 
   async getLibraryOrFail(): Promise<ILibrary> {
-    const library = await this.libraryRepository.findOne();
+    const library = await this.libraryRepository.findOne({});
```
Comment (hero101) [inline src/library/library/library.service.ts:27]: so... will this work?

File: src/platform/platfrom/platform.service.ts [hunk 15]
```diff
@@ -22,9 +22,11 @@ export class PlatformService {
   ) {}
 
   async getPlatformOrFail(
-    options?: FindOneOptions<Platform>
-  ): Promise<IPlatform> {
-    const platform = await this.platformRepository.findOne(options);
+    options: FindOneOptions<Platform>
```
Comment (hero101) [inline src/platform/platfrom/platform.service.ts]: why aren't the options optional?

File: src/platform/platfrom/platform.service.ts [hunk 16]
```diff
@@ -22,9 +22,11 @@ export class PlatformService {
...
+  ): Promise<IPlatform | never> {
+    let platform: IPlatform | null = null;
+    platform = (await this.platformRepository.find({ take: 1, ...options }))[0];
```
Comment (hero101) [inline src/platform/platfrom/platform.service.ts]:
```typescript
const platform = (await this.platformRepository.find({ take: 1, ...options }))?.[0];
```
will find return empty array of falsy?
the above code works both ways

File: src/platform/platfrom/platform.resolver.queries.ts [hunk 17]
```diff
@@ -16,7 +16,7 @@ export class PlatformResolverQueries {
     description: 'Alkemio Platform',
   })
   async platform(): Promise<IPlatform> {
-    return await this.platformService.getPlatformOrFail();
+    return await this.platformService.getPlatformOrFail({});
```
Comment (hero101) [inline src/platform/platfrom/platform.resolver.queries.ts]: why were the options changed to mandatory?

**Review by hero101 — CHANGES_REQUESTED** [review]
Let's go through the comments together because some of them are from lack of knowledge about what changed, others can be marked as won't fix.

File: src/domain/collaboration/agreement/agreement.entity.ts (hunk 4 above)
Comment (valentinyanakiev) [reply src/domain/collaboration/agreement/agreement.entity.ts:28]: The `name` column of `Tagset` is annotated with a default:

```typescript
  @Column({ default: RestrictedTagsetNames.DEFAULT })
  name: string;
```

It makes this call and parameter redundant.

File: src/domain/community/organization/organization.service.ts (hunk 10 above)
Comment (valentinyanakiev) [reply src/domain/community/organization/organization.service.ts]: this is duplicate with the one above

**Review by hero101 — APPROVED** [review]
